src/doc/trpl/dining-philosophers.md

```diff
@@ -0,0 +1,942 @@
+    ];
+
+    for p in &philosophers {
+        p.eat();
+    }
+
+    for _ in 0..5 {
+        done_rx.recv().unwrap();
+    }
+}
+```
+
+Just a few changes. Let’s break it down.
+
+```rust,ignore
+use std::thread;
+```
+
+We’ve added one more `use` line, as we’re going to start using the `thread`
+module.
+
+```rust,ignore
+    fn eat(&self) {
+        println!("{} is eating.", self.name);
+
+        thread::sleep_ms(1000);
+
+        self.done();
+    }
+```
+
+We have a new method, `eat()`. In it, we print out a message that our
+philosopher is eating, and then pause for a second. After we’ve taken
+that second to eat, we call `done()`. Easy enough!
+
+```rust,ignore
+    for p in &philosophers {
+        p.eat();
+    }
+```
+
+In `main()`, we now call `eat()` rather than `done()`.
+
+If you run this program, You should see each philosopher eat in turn:
+
+```text
+Baruch Spinoza is eating.
+Baruch Spinoza is done eating.
+Gilles Deleuze is eating.
+Gilles Deleuze is done eating.
+Karl Marx is eating.
+Karl Marx is done eating.
+Friedrich Nietzsche is eating.
+Friedrich Nietzsche is done eating.
+Michel Foucault is eating.
+Michel Foucault is done eating.
+```
+
+Excellent! We’re getting there. There’s just one problem: we aren’t actually
+operating in a concurrent fashion, which is a core part of the problem!
+
+To make our philosophers eat concurrently, we need to make a small change.
+Here’s the next iteration:
+
+```rust
+use std::sync::mpsc;
+use std::sync::mpsc::Sender;
+use std::thread;
+
+struct Philosopher {
+    name: String,
+    done: Sender<bool>
+}   
+
+impl Philosopher { 
+    fn new(name: &str, done: Sender<bool>) -> Philosopher {
+        Philosopher {
+            name: name.to_string(),
+            done: done,
+        }
+    }
+    
+    fn done(&self) {
+        println!("{} is done eating.", self.name);
+
+        self.done.send(true).ok().expect("Couldn't finish eating");
+    }
+
+    fn eat(&self) {
+        println!("{} is eating.", self.name);
+
+        thread::sleep_ms(1000);
+
+        self.done();
+    }
+}
+
+fn main() {
+    let (done_tx, done_rx) = mpsc::channel();
+
+    let philosophers = vec![
+        Philosopher::new("Baruch Spinoza", done_tx.clone()),
+        Philosopher::new("Gilles Deleuze", done_tx.clone()),
+        Philosopher::new("Karl Marx", done_tx.clone()),
+        Philosopher::new("Friedrich Nietzsche", done_tx.clone()),
+        Philosopher::new("Michel Foucault", done_tx.clone()),
+    ];
+
+    let handles: Vec<_> = philosophers.into_iter().map(|p| {
+        thread::spawn(move || {
+            p.eat();
+        })
+    }).collect();
+
+    for _ in 0..5 {
+        done_rx.recv().unwrap();
+    }
+
+    for h in handles {
+        h.join().ok().expect("Couldn't join a thread.");
+    }
+}
+```
+
+All we’ve done is change the loop in `main()`, and added a third one! Here’s the
+first change:
+
+```rust,ignore
+let handles: Vec<_> = philosophers.into_iter().map(|p| {
```

Comment:
I personally tend to prefer the type annotation here to be on the call to `collect` instead of the binding itself because the `collect` is all alone on its own line with tons of space to add this annotation, but this initial iterator construction is getting to be a bit of a longer line.